semantics/executable-spec/src/Control/Iterate/SetAlgebra.hs

```diff
@@ -137,22 +149,43 @@ data BiMap v a b where MkBiMap:: (v ~ b) => !(Map.Map a b) -> !(Map.Map b (Set.S
                                 --  ^   the 1st and 3rd parameter must be the same:             ^   ^
 
 -- ============== begin necessary Cardano.Binary instances ===============
-instance (Ord a, Ord b,ToCBOR a, ToCBOR b) => ToCBOR (BiMap b a b) where
-  toCBOR (MkBiMap l r) = encodeListLen 2 <> toCBOR l <> toCBOR r
-
-instance (Ord a, Ord b,FromCBOR a, FromCBOR b) => FromCBOR (BiMap b a b) where
-  fromCBOR = do
-    decodeListLenOf 2
-    !x <- fromCBOR
-    !y <- fromCBOR
-    return (MkBiMap x y)
+instance (Ord a, Ord b, ToCBOR a, ToCBOR b) => ToCBOR (BiMap b a b) where
+  -- The `toCBOR` instance encodes only the forward map. We wrap this in a
+  -- length-one list because a _previous_ encoding wrote out both maps, and we
+  -- can easily use the list length token to distinguish between them.
+  toCBOR (MkBiMap l _) = encodeListLen 1 <> toCBOR l
+
+instance
+  forall a b.
+  (Ord a, Ord b, FromCBOR a, FromCBOR b) =>
+  FromCBOR (BiMap b a b)
+  where
+  fromCBOR =
+    decodeListLen >>= \case
+      1 -> decodeMapAsBimap
+      -- Previous encoding of 'BiMap' encoded both the forward and reverse
+      -- directions. In this case we skip the reverse encoding. Note that,
+      -- further, the reverse encoding was from 'b' to 'a', not the current 'b'
+      -- to 'Set a', and hence the dropper reflects that.
+      2 -> do
+        !x <- decodeMapAsBimap
+        dropMap (void $ fromCBOR @b) (void $ fromCBOR @a)
+        return x
```

Comment:
I assume you have verified that the backwards compatibility indeed works as expected? :slightly_smiling_face:

Reply:
I've confirmed that this is able to decode the old snapshot.